libs/vkd3d/swapchain.c

```diff
@@ -2444,6 +2659,192 @@ static CONST_VTBL struct IDXGIVkSwapChainFactoryVtbl dxgi_vk_swap_chain_factory_
+{
+    return chain->present.low_latency_state.mode;
+}
+
+void dxgi_vk_swap_chain_latency_sleep(struct dxgi_vk_swap_chain* chain)
+{
+    const struct vkd3d_vk_device_procs *vk_procs = &chain->queue->device->vk_procs;
+    VkLatencySleepInfoNV latency_sleep_info;
+    VkSemaphoreWaitInfo sem_wait_info;
+    bool should_sleep = false;
+
+    /* Increment the low latency sem value before the wait */
+    chain->present.low_latency_sem_value++;
+
+    memset(&latency_sleep_info, 0, sizeof(latency_sleep_info));
+    latency_sleep_info.sType = VK_STRUCTURE_TYPE_LATENCY_SLEEP_INFO_NV;
+    latency_sleep_info.pNext = NULL;
+    latency_sleep_info.signalSemaphore = chain->present.low_latency_sem;
+    latency_sleep_info.value = chain->present.low_latency_sem_value;
+
+    pthread_mutex_lock(&chain->present.low_latency_swapchain_lock);
+
+    if (chain->present.vk_swapchain)
+    {
+        should_sleep = true;
+        VK_CALL(vkLatencySleepNV(chain->queue->device->vk_device, chain->present.vk_swapchain, &latency_sleep_info));
+    }
+
+    pthread_mutex_unlock(&chain->present.low_latency_swapchain_lock);
+
+    if (should_sleep)
+    {
+        memset(&sem_wait_info, 0, sizeof(sem_wait_info));
+        sem_wait_info.sType = VK_STRUCTURE_TYPE_SEMAPHORE_WAIT_INFO;
+        sem_wait_info.pNext = NULL;
+        sem_wait_info.flags = 0;
+        sem_wait_info.semaphoreCount = 1;
+        sem_wait_info.pSemaphores = &chain->present.low_latency_sem;
+        sem_wait_info.pValues = &chain->present.low_latency_sem_value;
+
+        VK_CALL(vkWaitSemaphores(chain->queue->device->vk_device, &sem_wait_info, UINT64_MAX));
+    }
+}
+
+void dxgi_vk_swap_chain_set_latency_sleep_mode(struct dxgi_vk_swap_chain* chain, bool low_latency_mode,
+	bool low_latency_boost, uint32_t minimum_interval_us)
+{
+    /* If the low latency state is already set to what the application is
+     * requesting, and there isn't an update in flight it is safe to skip
+     * this request. */
+    if (chain->low_latency_update_requested == false &&
+            low_latency_mode == chain->present.low_latency_state.mode &&
+            low_latency_boost == chain->present.low_latency_state.boost &&
+            minimum_interval_us == chain->present.low_latency_state.minimum_interval_us)
+        return;
+
+    pthread_mutex_lock(&chain->present.low_latency_state_update_lock);
+
+    chain->requested_low_latency_state.mode = low_latency_mode;
+    chain->requested_low_latency_state.boost = low_latency_boost;
+    chain->requested_low_latency_state.minimum_interval_us = minimum_interval_us;
+
+    /* The actual call to vkSetLatencySleepModeNV will happen
+     * in the present task when the swapchain is recreated. This
+     * is required when enabling low latency mode to use the frameId
+     * provided by the application, because the id it provides may be
+     * less than the largest id already used with the current swapchain. */
+    chain->low_latency_update_requested = true;
```

Comment:
It looks like some games in the wild spams SetSleepMode every frame, and recreating swapchain each time isn't going to fly unless swapchain recreation is super fast on Nvidia (I tested on AMD).

You can install Ghostrunner 2 (2144740) and use Reflex low latency with a frame limit also set in the game. The log will spam every frame:
```
NvAPI_D3D_SetSleepMode (Enabled/0us): OK
NvAPI_D3D_SetSleepMode (Enabled/6944us): OK
```
Note that the game's Reflex toggle is also buggy so make sure to change it to off then back to on again on each game restart.

Reply:
The check above this should eventually kick in to filter out redundant state updates. I will try that out though.

Reply:
The problem is that it's not exactly redundant. It's looping between having a minIntervalUs and not having it.

If I force minIntervalUs to be ignored and set to 0 then it seems to work.

Reply:
Ah I see that in your original message now, sorry about that. That does sound problematic, I will take a look. I can probably skip recreating the swapchain when the game is just setting the minimum interval.

Reply:
The latest revision works fine with the game. Thanks for the change.